src/server/master_server/chunk_server_manager.h

```diff
@@ -54,45 +137,132 @@ class ChunkServerManager {
 
   void operator=(const ChunkServerManager&) = delete;
 
-  // Allocates a ChunkServer to store a chunk.
-  // Returns nullptr if there's no available ChunkServer.
-  std::shared_ptr<protos::ChunkServer> AllocateChunkServer();
+  // This will allocate the specified number of chunk servers for storing the
+  // specified chunk. And will return the locations of the allocated servers.
+  // This is where we do some load balancing to make sure that the chunks are
+  // evenly distributed across chunkservers. We prioritize chunk servers with
+  // the most available disk space. This is threadsafe and allows concurrent
+  // allocations. Note:
+  // 1. The specified chunk_handle is now mapped to the returned locations.
+  // 2. Will return empty set of locations, if there is no chunk server
+  // available for chunk storage. e.g. When chunkservers don't have enough disk
+  // space left or there are no chunk servers registered (i.e. servers down).
+  // 3. If allocate is called for an already allocated chunk, we will return the
+  // previously allocated locations and not do new allocation.
```

Comment:
Yes you're correct. For that we will have to handle things a little bit different since we don't want to pick a chunkserver that already has that chunk again. So may need to create a diff method for that or something. But will leave this as it is for now, and when we get there, will think about the best way to handle that.

Reply:
will add a todo.